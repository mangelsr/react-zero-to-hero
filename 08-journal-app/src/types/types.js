export const types = {
    login: '[AUTH] LOGIN',
    logout: '[AUTH] LOGOUT',

    uiSetError: '[UI] SET ERROR',
    uiRemoveError: '[UI] REMOVE ERROR',
    
    uiStartLoading: '[UI] START LOADING',
    uiEndLoading: '[UI] END LOADING',

    notesAddNew: '[NOTES] NEW NOTE',
    notesActive: '[NOTES] SET ACTIVE NOTE',
    notesLoad: '[NOTES] LOAD NOTES',
    notesUpdate: '[NOTES] UPDATE NOTES',
    notesFile: '[NOTES] IMAGE NOTES',
    notesDelete: '[NOTES] DELETE NOTE',
    notesLogoutClean: '[NOTES] LOGOUT CLEAN',
};