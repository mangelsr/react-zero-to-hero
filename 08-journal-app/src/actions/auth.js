import { firebase, googleAuthProvider } from "../firebase/firebaseConfig";
import Swal from "sweetalert2";

import { types } from "../types/types";
import { uiEndLoading, uiStartLoading } from "./ui";
import { clearNotes } from "./notes";

export const startLoginEmailPassword = (email, password) => {
    return async (dispatch) => {
        try {
            dispatch(uiStartLoading());
            const { user } = await firebase.auth().signInWithEmailAndPassword(email, password);
            dispatch(login(user.uid, user.displayName));
        } catch (e) {
            console.log(e);
            if (e.message) {
                Swal.fire('Error', e.message, 'error');
            }
        } finally {
            dispatch(uiEndLoading());
        }
    };
}

export const startGoogleLogin = () => {
    return async (dispatch) => {
        try {
            dispatch(uiStartLoading());
            const { user } = await firebase.auth().signInWithPopup(googleAuthProvider);
            dispatch(login(user.uid, user.displayName))
        } catch (e) {
            console.log(e);
            if (e.message) {
                Swal.fire('Error', e.message, 'error');
            }
        } finally {
            dispatch(uiEndLoading());
        }
    };
};


export const startRegister = (email, password, name) => {
    return async (dispatch) => {
        try {
            dispatch(uiStartLoading());
            const { user } = await firebase.auth().createUserWithEmailAndPassword(email, password);
            await user.updateProfile({displayName: name});
            dispatch(login(user.uid, user.displayName));
        } catch (e) {
            console.log(e);
            if (e.message) {
                Swal.fire('Error', e.message, 'error');
            }
        } finally {
            dispatch(uiEndLoading());
        }
    };
};

export const login = (uid, displayName) => ({
    type: types.login,
    payload: {
        uid,
        displayName
    }
})

export const startLogout = () => {
    return async (dispatch) => {
        try {
            await firebase.auth().signOut();
            dispatch(logout());
            dispatch(clearNotes());
        } catch (e) {
            console.log(e);
            if (e.message) {
                Swal.fire('Error', e.message, 'error');
            }
        }
    };
};

export const logout = () => ({
    type: types.logout
});