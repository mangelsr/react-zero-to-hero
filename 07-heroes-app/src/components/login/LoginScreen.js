import React, { useContext } from 'react';
import { AuthContext } from '../../auth/AuthContext';
import { types } from '../../types/types';


export const LoginScreen = ({ history }) => {

    const { dispatch } = useContext(AuthContext);
    
    const handleLogin = () => {
        dispatch({
            type: types.login,
            payload: {
                name: 'Miguel'
            }
        });
        const lastPath = localStorage.getItem('lastpath') || '/';
        history.replace(lastPath);
    };

    return (
        <div className="container mt-5">
            <h1>Login Screen</h1>
            <hr/>
            <button
                onClick={handleLogin}
                className="btn btn-outline-success"
                >
                Login
            </button>
        </div>
    )
}
