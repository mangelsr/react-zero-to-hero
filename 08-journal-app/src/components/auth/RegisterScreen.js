import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import validator from "validator";

import { startRegister } from '../../actions/auth';
import { removeError, setError } from '../../actions/ui';
import { useForm } from '../../hooks/useForm';


export const RegisterScreen = () => {

    const dispatch = useDispatch();
    const { loading, msgError } = useSelector(state => state.ui);

    const [ formValues, handleInputChange ] = useForm({
        name: 'Angel',
        email: 'angel@gmail.com',
        password: '123456',
        password2: '123456',
    });

    const { name, email, password, password2 } = formValues;

    const handleRegister = e => {
        e.preventDefault();
        if (isFormValid()) {
            dispatch(startRegister(email, password, name));
        }
    };

    const isFormValid = () => {
        if (name.trim().length === 0) {
            dispatch(setError('Name is required'));
            return false;
        } else if (!validator.isEmail(email)) {
            dispatch(setError('Email is not valid'));
            return false;
        } else if (password !== password2 || password.length < 5) {
            dispatch(setError('Password should match and be at least 6 characers'));
            return false;
        }
        dispatch(removeError());
        return true;
    };

    return (
        <div className="animate__animated animate__fadeIn">
            <h3 className="auth__title">Register</h3>
            <form onSubmit ={handleRegister}>

                { msgError && 
                    <div className="auth__alert-error">
                        { msgError }
                    </div>
                }
                
                <input onChange={handleInputChange} value={name} className="auth__input" type="text" placeholder="Name" name="name" autoComplete="off"/>
                <input onChange={handleInputChange} value={email} className="auth__input" type="text" placeholder="Email" name="email" autoComplete="off"/>
                <input onChange={handleInputChange} value={password} className="auth__input" type="password" placeholder="Password" name="password"/>
                <input onChange={handleInputChange} value={password2} className="auth__input" type="password" placeholder="Password confirmation" name="password2"/>
                
                <button disabled={loading} className="btn btn-primary btn-block mb-5" type="submit">Register</button>

                <Link className="link" to="/auth/login">
                    Already registered?
                </Link>
            </form>
        </div>
    );
}
