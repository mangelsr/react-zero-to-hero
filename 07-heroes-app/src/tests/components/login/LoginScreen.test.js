import '@testing-library/jest-dom';
import { mount } from 'enzyme';
import React from 'react';
import { AuthContext } from '../../../auth/AuthContext';
import { LoginScreen } from '../../../components/login/LoginScreen';
import { types } from '../../../types/types';


describe('Tests over <LoginScreen/> component', () => {

    const historyMock = {
        replace: jest.fn(),
    };

    const userContext = {
        dispatch: jest.fn(),
    };

    const wrapper = mount(
        <AuthContext.Provider value={userContext}>
            <LoginScreen history={historyMock} />
        </AuthContext.Provider>
    );

    test('should render the component', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('should dispatch action and called replace', () => {
        wrapper.find('button').simulate('click');
        
        expect(userContext.dispatch).toHaveBeenCalledTimes(1);
        expect(userContext.dispatch).toHaveBeenCalledWith({
            type: types.login,
            payload: {
                name: 'Miguel'
            }
        });

        expect(historyMock.replace).toHaveBeenCalled();
        expect(historyMock.replace).toHaveBeenCalledWith('/');

        const lastPath = '/dc';
        localStorage.setItem('lastpath', lastPath);
        
        wrapper.find('button').simulate('click');
        
        expect(historyMock.replace).toHaveBeenCalledWith(lastPath);
    });
    
});