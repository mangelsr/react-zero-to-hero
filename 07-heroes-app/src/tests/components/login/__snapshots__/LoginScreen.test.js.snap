// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests over <LoginScreen/> component should render the component 1`] = `
<div
  className="container mt-5"
>
  <h1>
    Login Screen
  </h1>
  <hr />
  <button
    className="btn btn-outline-success"
    onClick={[Function]}
  >
    Login
  </button>
</div>
`;
